### PR TITLE
[RFC] vim-patch:8.1.{729,732}

### LINKED ISSUE #3739: packadd autocmd
I would like to be able to run some code after a plugin was loaded. One example use-case would be call `neomake#configure#automake('nrwi', 500)` which is necessary to configure the neomake plugin. Currently it is hard to do that when neomake is an optional plugin.

I imagine something like this:

```
autocmd PackLoadPre neomake packadd some-dependency
autocmd PackLoadPost neomake call neomake#configure#automake('nrwi', 500)
```

The autocommand should only be triggered on the first `:packadd`. The natural place to call these autocommands seems to be the `add_pack_dir_to_rtp` function:
https://github.com/vim/vim/blob/af0839acec403379627a4bc50cbeca588344cea5/src/ex_cmds2.c#L3684

I previously opened https://github.com/neovim/neovim/issues/9415 for this, where I was redirected here. It was suggested to use `SourcePre`, which should be able to replace most usecases but would more cumbersome to use correctly.

### PR DESCRIPTION
**Pull request by janlazo**
**vim-patch:8.1.0729: there is a SourcePre autocommand event but not a SourcePost**
Problem:    There is a SourcePre autocommand event but not a SourcePost.
Solution:   Add the SourcePost autocommand event. (closes vim/vim#3739)
vim/vim@2b61852

Leaving this WIP until patches for ~~VIM_CLEAR~~ and sourcing files are merged.

### CODE REVIEW

**Comment by justinmk** [comment]
```
==15691==ERROR: AddressSanitizer: heap-use-after-free on address 0x616000002c48 at pc 0x000000c4642d bp 0x7fff5dda60f0 sp 0x7fff5dda60e8
READ of size 8 at 0x616000002c48 thread T0
    #0 0xc4642c in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3295:42
    #1 0xc40bd4 in cmd_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2977:14
    #2 0xc41390 in ex_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2958:3
    #3 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #4 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #5 0x932ecc in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21905:3
    #6 0x8fcb7b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6392:11
    #7 0x9122f3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6149:11
    #8 0x90ae7e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2744:9
    #9 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #10 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #11 0x95472a in ex_execute /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:20027:7
    #12 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #13 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #14 0x932ecc in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21905:3
    #15 0x8fcb7b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6392:11
    #16 0x9122f3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6149:11
    #17 0x90ae7e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2744:9
    #18 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #19 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #20 0xc44e22 in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3237:3
    #21 0xc40bd4 in cmd_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2977:14
    #22 0xc41390 in ex_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2958:3
    #23 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #24 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #25 0xc659a5 in do_cmdline_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:269:10
    #26 0xff54ab in exe_commands /home/travis/build/neovim/neovim/build/../src/nvim/main.c:1706:5
    #27 0xfe2811 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:548:5
    #28 0x7f9cafd5982f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #29 0x450ab8 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x450ab8)

0x616000002c48 is located 456 bytes inside of 608-byte region [0x616000002a80,0x616000002ce0)
freed by thread T0 here:
    #0 0x509548 in realloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x509548)
    #1 0x111623c in xrealloc /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:151:15
    #2 0xea2d4a in ga_grow /home/travis/build/neovim/neovim/build/../src/nvim/garray.c:98:14
    #3 0xc43557 in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3185:5
    #4 0xc3f5fe in source_callback /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2372:9
    #5 0xc3eeb4 in do_in_path /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2450:15
    #6 0xc3f2bb in do_in_path_and_pp /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2494:12
    #7 0xc3f5d1 in source_in_path /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2542:10
    #8 0xc3d7dc in source_runtime /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2536:10
    #9 0x92c597 in script_autoload /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21310:9
    #10 0x919245 in find_var_in_ht /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:19207:12
    #11 0x94b427 in find_var /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:19149:27
    #12 0x98e9c5 in get_var_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:18946:7
    #13 0x9a1073 in eval7 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:4271:15
    #14 0x99d033 in eval6 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3944:7
    #15 0x99aefa in eval5 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3825:9
    #16 0x995c3f in eval4 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3532:7
    #17 0x99511c in eval3 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3450:7
    #18 0x9945ec in eval2 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3381:7
    #19 0x8f5672 in eval1 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3308:7
    #20 0x953267 in ex_execute /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:19979:9
    #21 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #22 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #23 0xc44e22 in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3237:3
    #24 0xc40bd4 in cmd_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2977:14
    #25 0xc41390 in ex_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2958:3
    #26 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #27 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #28 0x932ecc in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21905:3
    #29 0x8fcb7b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6392:11

previously allocated by thread T0 here:
    #0 0x509548 in realloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x509548)
    #1 0x111623c in xrealloc /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:151:15
    #2 0xea2d4a in ga_grow /home/travis/build/neovim/neovim/build/../src/nvim/garray.c:98:14
    #3 0xc43557 in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3185:5
    #4 0xc40bd4 in cmd_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2977:14
    #5 0xc41390 in ex_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2958:3
    #6 0xc81050 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2243:5
    #7 0xc5f7d7 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #8 0xc659a5 in do_cmdline_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:269:10
    #9 0xff54ab in exe_commands /home/travis/build/neovim/neovim/build/../src/nvim/main.c:1706:5
    #10 0xfe2811 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:548:5
    #11 0x7f9cafd5982f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
```

**Comment by janlazo** [comment]
Did it get worse after `XFREE_CLEAR`?

**Comment by justinmk** [comment]
> Did it get worse after `XFREE_CLEAR`?

Can check that by removing `*ptr = NULL` line in `XFREE_CLEAR` macro.

**Comment by justinmk** [comment]
vim-patch 8.1.0362 adds `current_sctx`. Maybe that or related patches changed the codepaths somehow?

**Comment by blueyed** [comment]
8.1.1365 is included in https://github.com/neovim/neovim/pull/10309 now.

File: src/nvim/ex_cmds2.c [hunk 1]
```diff
@@ -3181,7 +3186,7 @@ int do_source(char_u *fname, int check_other, int is_vimrc)
     }
     si = &SCRIPT_ITEM(current_SID);
     si->sn_name = fname_exp;
-    fname_exp = NULL;
+    fname_exp = vim_strsave(si->sn_name);  // used for autocmd
```
Comment (janlazo) [inline src/nvim/ex_cmds2.c]: Why is this enough to pass the build?

Comment (justinmk) [reply src/nvim/ex_cmds2.c]: `si` is part of `script_items`. When `ga_grow(&script_items,...)` reallocates (see stacktrace above), `si->sn_name` pointer is invalid. Creating a copy here avoids that.

**Pull request merged**